zcash_client_sqlite/src/lib.rs

```diff
@@ -396,6 +402,30 @@ impl<'a, P: consensus::Parameters> DataConnStmtCache<'a, P> {
 
 #[allow(deprecated)]
 impl<'a, P: consensus::Parameters> WalletWrite for DataConnStmtCache<'a, P> {
+    fn get_next_available_address(
+        &mut self,
+        account: AccountId,
+    ) -> Result<Option<UnifiedAddress>, Self::Error> {
+        match self.get_unified_full_viewing_keys()?.get(&account) {
+            Some(ufvk) => {
+                let mut search_from = wallet::get_current_address(self.wallet_db, account)?
+                    .map_or_else(DiversifierIndex::default, |(_, last_diversifier_index)| {
+                        last_diversifier_index
+                    });
+                search_from
+                    .increment()
+                    .map_err(|_| SqliteClientError::DiversifierIndexOutOfRange)?;
```

Comment:
This is subtly incorrect, in a way that I believe doesn't actually break:

- If `wallet::get_current_address()` returns `Ok(Some(_)`, that means we've generated addresses for this account before, and we start searching from `last_diversifier_index + 1`, as expected.
- If `wallet::get_current_address()` returns `Ok(None)`, that means there is no entry in the `addresses` table for the given account. That implies no address has ever been generated for the account. In this scenario, we should start searching from `DiversifierIndex::zero()`, but we actually start from `DiversifierIndex::zero() + 1`, which skips an address.

In practice this doesn't break because when adding an account, we also always add the default UA, so we _never_ get `Ok(None)`. However, we should for correctness fix this:
```suggestion
                let search_from = wallet::get_current_address(self.wallet_db, account)?
                    .map(|(_, last_diversifier_index)| {
                        last_diversifier_index
                            .increment()
                            .map_err(|_| SqliteClientError::DiversifierIndexOutOfRange)
                    })
                    .transpose()?
                    .unwrap_or_else(DiversifierIndex::default);
```

Reply:
This doesn't work, because `increment` does mutation in place, so the `map` here results in `Option<()>`. But I agree that the increment is in the wrong place.